File: le-spark/src/main/resources/scripts/initialize.scala
import com.fasterxml.jackson.databind.node.ObjectNode
import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
import com.fasterxml.jackson.module.scala.DefaultScalaModule
import com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper
import org.apache.spark.sql.{DataFrame, SparkSession}

import scala.collection.JavaConverters._

class LatticeContext(val input: List[DataFrame], val params: JsonNode, val targets: List[ObjectNode]) {
  var output: List[DataFrame] = List[DataFrame]()
  var outputStr: String = ""
}

val mapper = new ObjectMapper() with ScalaObjectMapper
mapper.registerModule(DefaultScalaModule)

val spark = SparkSession.builder().appName("SparkSession").getOrCreate()

val checkpointDir = """{{CHECKPOINT_DIR}}"""

if (checkpointDir.length > 0) {
  spark.sparkContext.setCheckpointDir(checkpointDir)
  println("----- BEGIN SCRIPT OUTPUT -----")
  println(s"Checkpoint Dir: $checkpointDir")
  println("----- END SCRIPT OUTPUT -----")
}

val rawInput = mapper.readValue[List[JsonNode]]("""{{INPUT}}""")

def loadHdfsUnit(unit: JsonNode): DataFrame = {
  var path = unit.get("Path").asText()
  val fmt = if (unit.get("DataFormat") != null) unit.get("DataFormat").asText.toLowerCase else "avro"
  val partitionKeys = if (unit.get("PartitionKeys") == null) List() //
  else unit.get("PartitionKeys").elements.asScala.map(_.asText()).toList
  if (partitionKeys.isEmpty) {
    val suffix = "." + fmt
    if (!path.endsWith(suffix)) {
      if (path.endsWith("/")) {
        path += "*" + suffix
      } else {
        path += "/*" + suffix
      }
    }
  }
  spark.read.format(fmt).load("hdfs://" + path)
}

val scriptInput = rawInput map { input => {
  val storage = input.get("StorageType").asText().toLowerCase()
  storage match {
    case "hdfs" => loadHdfsUnit(input)
    case _ => throw new UnsupportedOperationException(s"Unknown storage $storage")
  }
}
}

println("----- BEGIN SCRIPT OUTPUT -----")
println(s"Input: $scriptInput")
println("----- END SCRIPT OUTPUT -----")

val scriptParams = mapper.readValue[JsonNode]("""{{PARAMS}}""")

println("----- BEGIN SCRIPT OUTPUT -----")
println(s"Params: $scriptParams")
println("----- END SCRIPT OUTPUT -----")

val scriptTargets = mapper.readValue[List[ObjectNode]]("""{{TARGETS}}""")

println("----- BEGIN SCRIPT OUTPUT -----")
println(s"Targets: $scriptTargets")
println("----- END SCRIPT OUTPUT -----")

val lattice: LatticeContext = new LatticeContext(scriptInput, scriptParams, scriptTargets)

def setPartitionTargets(index: Int, partitionKeys: Seq[String], lattice: LatticeContext): Unit = {
  if (index >= 0 && index < lattice.targets.size) {
    lattice.targets(index).set("PartitionKeys", mapper.valueToTree(partitionKeys))
  } else {
    throw new RuntimeException(s"There's no Target $index")
  }
}

File: le-spark/src/main/resources/scripts/initialize.py
import json


class LatticeContext:
    def __init__(self, input, params, targets):
        self.input = input
        self.params = params
        self.targets = targets
        self.output = []
        self.output_str = ""


def load_data_unit(unit):
    storage = unit['StorageType'].lower()
    if storage == "hdfs":
        return load_hdfs_unit(unit)
    else:
        raise ValueError("Unsupported storage type %s" % storage)


def load_hdfs_unit(unit):
    path = unit['Path']
    fmt = unit['DataFormat'] if 'DataFormat' in unit else "avro"
    partition_keys = unit['PartitionKeys'] if 'PartitionKeys' in unit else []
    if (partition_keys is None) or (len(partition_keys) == 0):
        suffix = "." + fmt
        if path[-len(suffix):] != suffix:
            if path[-1] == "/":
                path += "*" + suffix
            else:
                path += "/*" + suffix
    path = "hdfs://%s" % path
    return spark.read.format(fmt).load(path)


checkpoint_dir = '''{{CHECKPOINT_DIR}}'''
if checkpoint_dir != "":
    print("----- BEGIN SCRIPT OUTPUT -----")
    print("Checkpoint Dir:", checkpoint_dir)
    print("----- END SCRIPT OUTPUT -----")
    spark.sparkContext.setCheckpointDir(checkpoint_dir)

script_targets = json.loads('''{{TARGETS}}''')
print("----- BEGIN SCRIPT OUTPUT -----")
print("Targets:", script_targets)
print("----- END SCRIPT OUTPUT -----")

raw_input = json.loads('''{{INPUT}}''')
script_input = [load_data_unit(unit) for unit in raw_input]

print("----- BEGIN SCRIPT OUTPUT -----")
print("Input:", script_input)
print("----- END SCRIPT OUTPUT -----")

script_params = json.loads('''{{PARAMS}}''')

print("----- BEGIN SCRIPT OUTPUT -----")
print("Params: %s" % json.dumps(script_params))

lattice = LatticeContext(input=script_input, params=script_params, targets=script_targets)


def set_partition_targets(index, lst, lattice):
    if (index >= 0) and (index < len(lattice.targets)):
        lattice.targets[index]['PartitionKeys'] = lst
    else:
        raise Exception("Index not exist %d" % index)
